ValveResourceFormat/Resource/ResourceTypes/KV3.cs

```diff
@@ -0,0 +1,627 @@
+        public string Key;
+        public Dictionary<String, KVValue> Properties;
+        private bool IsArray;
+        public int Count;
+
+        public KVObject(string name)
+        {
+            Key = name;
+            Properties = new Dictionary<String, KVValue>();
+            IsArray = false;
+            Count = 0;
+        }
+
+        public KVObject(string name, bool isArray)
+            : this(name)
+        {
+            IsArray = isArray;
+        }
+
+        //Add a property to the structure
+        public virtual void AddProperty(String name, KVValue value)
+        {
+            if (IsArray)
+            {
+                //Make up a key for the dictionary
+                Properties.Add(Count.ToString(), value);
+            }
+            else
+            {
+                Properties.Add(name, value);
+            }
+
+            Count++;
+        }
+
+        public void Serialize(StringBuilder stringBuilder, int indent)
+        {
+            if (IsArray)
+            {
+                SerializeArray(stringBuilder, indent);
+            }
+            else
+            {
+                SerializeObject(stringBuilder, indent);
+            }
+        }
+
+        //Serialize the contents of the KV object
+        private void SerializeObject(StringBuilder stringBuilder, int indent)
+        {
+            stringBuilder.Append(Environment.NewLine);
+            for (var i = 0; i < indent - 1; i++)
+            {
+                stringBuilder.Append("\t");
+            }
+            stringBuilder.Append("{");
+            stringBuilder.Append(Environment.NewLine);
+
+            foreach (var pair in Properties)
+            {
+                for (var i = 0; i < indent; i++)
+                {
+                    stringBuilder.Append("\t");
+                }
+
+                stringBuilder.Append(pair.Key);
+                stringBuilder.Append(" = ");
+
+                PrintValue(stringBuilder, pair.Value.Type, pair.Value.Value, indent);
+
+                stringBuilder.Append(Environment.NewLine);
+            }
+
+            for (var i = 0; i < indent - 1; i++)
+            {
+                stringBuilder.Append("\t");
+            }
+            stringBuilder.Append("}");
+        }
+
+        private void SerializeArray(StringBuilder stringBuilder, int indent)
+        {
+            stringBuilder.Append(Environment.NewLine);
+            for (var i = 0; i < indent - 1; i++)
+            {
+                stringBuilder.Append("\t");
+            }
+            stringBuilder.Append("[");
+            stringBuilder.Append(Environment.NewLine);
+
+            //Need to preserve the order
+            for (var i = 0; i < Count; i++)
+            {
+                for (var j = 0; j < indent; j++)
+                {
+                    stringBuilder.Append("\t");
+                }
+
+                PrintValue(stringBuilder, Properties[i.ToString()].Type, Properties[i.ToString()].Value, indent);
```

Comment:
Why are you splitting KVValue up into its two fields rather than just sending the one in?